test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp

```diff
@@ -257,3 +325,75 @@ TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNonBlocking) {
 
   // Communicators might be aborted here, further operations would fail.
 }
+
+TEST_F(ProcessGroupNCCLErrorsTest, testNCCLErrorsNoHeartbeat) {
+  if (skipTest()) {
+    return;
+  }
+
+  int heartBeatIntervalInSec = 2;
+  std::string timeInterval = std::to_string(heartBeatIntervalInSec);
+  ASSERT_TRUE(setenv(c10d::NCCL_BLOCKING_WAIT, "1", 1) == 0);
+  ASSERT_TRUE(
+      setenv(c10d::TORCH_NCCL_HEARTBEAT_TIMEOUT_S, timeInterval.c_str(), 1) ==
+      0);
+  ASSERT_TRUE(setenv(c10d::TORCH_NCCL_ENABLE_MONITORING, "1", 1) == 0);
+  auto options = c10d::ProcessGroupNCCL::Options::create();
+  // Set a long watchdog timeout, so that we have enough time to lock the
+  // watchdog and let the heartbeat monitor thread to kick in.
+  options->timeout = std::chrono::milliseconds(30000);
+  ProcessGroupNCCLNoHeartbeat pg(store_, 0, 1, options);
+
+  // Normal collective case.
+  auto work = pg.allreduce(tensors_);
+  work->wait();
+  EXPECT_TRUE(work->isSuccess());
+
+  work = pg.allreduce(tensors_);
+  {
+    // Now run all reduce with errors.
+    std::lock_guard<std::mutex> lock(pg.getWatchdogMutex());
+    LOG(INFO) << "Lock watchdog thread.";
+    // Wait long enough before monitor thread throws exceptions.
+    std::this_thread::sleep_for(
+        std::chrono::seconds(heartBeatIntervalInSec * 3));
+    // Check the monitoring thread launched and exception thrown.
+    EXPECT_TRUE(pg.getErrorCaughtFlag());
+  }
+  work->wait();
+  EXPECT_TRUE(work->isSuccess());
+}
+
+TEST_F(ProcessGroupNCCLErrorsTest, testNCCLTimedoutDebugInfo) {
+  if (skipTest()) {
+    return;
+  }
+
+  int heartBeatIntervalInSec = 3;
+  std::string timeInterval = std::to_string(heartBeatIntervalInSec);
+  ASSERT_TRUE(setenv(c10d::NCCL_BLOCKING_WAIT, "1", 1) == 0);
+  ASSERT_TRUE(
+      setenv(c10d::TORCH_NCCL_HEARTBEAT_TIMEOUT_S, timeInterval.c_str(), 1) ==
+      0);
+  ASSERT_TRUE(setenv(c10d::TORCH_NCCL_ENABLE_MONITORING, "1", 1) == 0);
+  ASSERT_TRUE(setenv(c10d::NCCL_DESYNC_DEBUG, "1", 1) == 0);
+  // We cannot capture the exception thrown in watchdog thread without making
+  // lots of changes to the code. So we don't let the watchdog throw exception.
+  ASSERT_TRUE(setenv(c10d::NCCL_ASYNC_ERROR_HANDLING, "0", 1) == 0);
```

Comment:
didn't understand why we have to set async = 0 to do this test.

what if we leave async=1, then let a nccl timeout happen.  (issue collective on rank0 not rank1).  then we 'observe' that desync report gets run (bc we made a class function that we can override in the test).  then we check that heartbeat monitor thread shuts down cleanly (without firing).

Reply:
then also, we can add an alternate version of this test where everything is the same as above, except we make desyncDebug function hang for too long (sleep).  We test that heartbeat monitor fires and kills in this case.

Reply:
setting `NCCL_ASYNC_ERROR_HANDLING ` to 0 is to avoid watchdog throw exception and get the whole process aborted which cannot be try-catched. I can add logic of whether monitor filed or not.. And also will add one more test case which mentioned in your second comment.

Reply:
ok i think this is fine.